tests/system/framework/hosts/shadow.py

```diff
@@ -0,0 +1,128 @@
+"""shadow multihost host."""
+
+from __future__ import annotations
+
+from pathlib import PurePosixPath
+from typing import Any
+
+from pytest_mh.conn import ProcessLogLevel
+
+from .base import BaseHost, BaseLinuxHost
+
+__all__ = [
+    "ShadowHost",
+]
+
+
+class ShadowHost(BaseHost, BaseLinuxHost):
+    """
+    shadow host object.
+
+    This is the host where the tests are run.
+
+    .. note::
+
+        Full backup and restore of shadow state is supported.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self._features: dict[str, bool] | None = None
+
+    def pytest_setup(self) -> None:
+        super().pytest_setup()
+
+    def start(self) -> None:
+        """
+        Not supported.
+
+        :raises NotImplementedError: _description_
+        """
+        raise NotImplementedError("Starting shadow service is not implemented.")
+
+    def stop(self) -> None:
+        """
+        Not supported.
+
+        :raises NotImplementedError: _description_
+        """
+        raise NotImplementedError("Stopping shadow service is not implemented.")
+
+    def backup(self) -> Any:
+        """
+        Backup all shadow data.
+
+        :return: Backup data.
+        :rtype: Any
+        """
+        self.logger.info("Creating backup of shadow host")
+
+        result = self.conn.run(
+            """
+            set -ex
+
+            function backup {
+                if [ -d "$1" ] || [ -f "$1" ]; then
+                    cp --force --archive "$1" "$2"
+                fi
+            }
+
+            path=`mktemp -d`
+            backup /etc/login.defs "$path/login.defs"
+            backup /etc/default/useradd "$path/useradd"
+            backup /etc/passwd "$path/passwd"
+            backup /etc/shadow "$path/shadow"
+            backup /etc/group "$path/group"
+            backup /etc/gshadow "$path/gshadow"
+            backup /etc/subuid "$path/subuid"
+            backup /etc/subgid "$path/subgid"
+            backup /var/log/secure "$path/secure"
+
+            echo $path
+            """,
+            log_level=ProcessLogLevel.Error,
+        )
+
+        return PurePosixPath(result.stdout_lines[-1].strip())
```

Comment:
Hm, but the script does set -e, so result.stdout_lines could in theory be empty?

I guess it'll just give us an indexerror so no big deal?  However, catching a return value != 0 early might be less confusing

Reply:
How about the following?
```
if not result.stdout_lines:
    return PurePosixPath(result.stdout_lines[-1].strip())
else:
    return -1
```

Reply:
It shouldn't be empty, if the script fails it raises `ProcessError` from `run`  since `raise_on_error=True` by default.